README
www/index.txt

ChangeLog
2006-09-21  Karl Chen  <quarl@quarl.org>
    * Release v1.4.1
    * Hosted at Cubewano.org

2006-01-20  Karl Chen  <quarl@quarl.org>
    * Added support for EXIF ImageDescription (enable with DISPLAY_INFO)

    * New template/stylesheet/etc structure: quickypix/src/client is a
      directory containing templates/ and skin/.  Skin/ contains files
      served statically to the client and may be Aliased through Apache
      for efficiency.  Factored the various stylesheets; quickypix.css
      contains the common definitions.  Begin working on a new style that
      fixes image position.

    * Use first Pic as default highlight instead of first item (Pic or
      Album); reported by Grant Brown

    * Fixed crash when highlight pointed to an empty album; reported by
      Grant Brown

    * Don't write default values for .title, .highlight to disk (added
      "default" feature to MetadataProperty); reported by Grant Brown

    * Add configuration option for whether/how to sort root and non-root
      albums (reported by Grant Brown)

2006-01-19  Karl Chen  <quarl@quarl.org>
    * Removed '&' from banned filename characters, issue reported by Grant
      Brown

2005-12-16  Karl Chen  <quarl@quarl.org>
    * Created quickypix.cgi.su, a setuid wrapper (written in C, based on
      Apache 2.0.54's suexec).
    * New recommended permissions setup (install.html)
    * Added configuration option ALWAYS_USE_PUBLIC_FOR_PHOTOS: in admin
      mode, all jpgs are linked to the public root.  Advantages:
        - Client only needs to download pictures once if you move between
          roots
        - If the edit root and public root (s)cgi processes are under
          different userids, permissions for the cache directory are
          easier to manage if only one userid writes to it.
    * Factored quickypix.py to Presenter.py and quickypix.cgi
    * Move conversion scripts to new directory src/import/

2005-12-15  Karl Chen  <quarl@quarl.org>
    * Release v1.3.0

2005-12-14  Karl Chen  <quarl@quarl.org>
    * Added convert_gallery_dat (which was written 2005-01-12 and
      previously in the 'quarl' repository)

2005-12-07  Karl Chen  <quarl@quarl.org>
    * Make path configuration more robust

    * By default auto-detect the src/helpers/bin directory instead of
      requiring configuring it manually

    * Support older versions of ImageMagick that don't have '-thumbnail'
      option, via configuration parameter USE_IMAGEMAGICK_THUMBNAIL_OPTION

2005-12-06  Karl Chen  <quarl@quarl.org>
    * Fixed bug with caching thumbnails at root of image repository.  Bug
      reported by Grant Brown <grant@marstechnologies.com>.

2005-12-05  Karl Chen  <quarl@quarl.org>
    * Add tdel, movie2jpg, movie-rotate helpers to QuickyPix repository
      under quickypix/src/helpers/bin [originally at
      http://svn.quarl.org/repos/quarl/trunk/bin]
    * Add configuration options to set the location of various helper
      programs (the newly distributed ones as well as ImageMagick,
      Mplayer, Transcode). 
      
2005-12-05  Karl Chen  <quarl@quarl.org>
    * By default prefer ImageMagick to PIL (set USE_PIL=True to change).
    * PIL support: enable anti-aliasing and set JPEG quality 90.  TODO:
      adaptively set the JPEG quality based on input quality, the way
      ImageMagick does.
    * Fixed bug where comments were read and saved from 'name.comment'
      instead of 'name.jpg.comment'.  (Not really a bug, but inconsistent
      with the documentation; 'name.comment' behavior was accidentally
      introduced with the drop of extensions in URLs.)  Bug reported by
      Grant Brown <grant@marstechnologies.com>.

2005-12-04  Karl Chen  <quarl@quarl.org>
    * Added convert_descript_ion script for converting from 4DOS/ACDSee
      descript.ion files

2005-12-01  Karl Chen  <quarl@quarl.org>
    * Allow non-lowercase extensions for *.jpeg, etc.

2005-11-29  Karl Chen  <quarl@quarl.org>
    * Add helpful message when invoking quickypix.py on command-line
    * Reorganized some config setup
    * Now compatible with python2.2
      
2005-11-28  Karl Chen  <quarl@quarl.org>
    * Installation documentation additions; README for src/html

2005-11-02  Karl Chen  <quarl@quarl.org>
    * Release v1.2.2

    * Use PRAT [http://www.quarl.org/projects/prat/]

2005-11-01  Karl Chen  <quarl@quarl.org>
    * Fixed form problem in SCGI mode

2005-10-21  Karl Chen  <quarl@quarl.org>
    * Modularize info box lines; configure which ones to display

    * Configure default album title; default to '%(basename)s' instead
      of '(%(basename)s)'

    * Updated installation notes

2005-10-21  Hollis Blanchard  <hollis@penguinppc.org>
    * Info box: show exposure, shutter speed, and ISO rating

    * Updated installation notes for lighttpd

2005-10-15  Karl Chen  <quarl@quarl.org>
    * Fix another 'value.recent' bug

2005-10-15  Hollis Blanchard  <hollis@penguinppc.org>
    * Fix SCGI bug to prevent quickypix from exiting after every
      invocation

2005-10-15  Karl Chen  <quarl@quarl.org>
    * Added missing copyright statements
    * Use explicit python in init.d script

2005-10-15  Karl Chen  <quarl@quarl.org>
    * Release v1.2.1 (brown paper bag - fixed get_ext bug)

2005-10-15  Karl Chen  <quarl@quarl.org>
    * Release v1.2

2005-10-15  Karl Chen  <quarl@quarl.org>
    * Changed URL for fullsize pics to IMAGE.jpg instead of
      IMAGE/fullsize (resized pics are still IMAGE/123x456)

2005-10-15  Karl Chen  <quarl@quarl.org>
    * src/quickypix/util.py (MetadataProperty.set): Fixed 'recent' bug
      due to 2005-09-26 change
    * Other minor bug fixes

2005-10-14  Karl Chen  <quarl@quarl.org>
    * init.d/quickypix-scgi: startup script

2005-10-13  Hollis Blanchard  <hollis@penguinppc.org>
    * New SCGI handler.

2005-10-12  Karl Chen  <quarl@quarl.org>
    * Based on patch by Hollis; move config.root setting to config.py
      and add comments for alternate ways to set config.root.


src/quickypix/util.py
# $Id: util.py 325 2006-01-21 02:25:11Z quarl $

## Copyright (C) 2005, 2006 Karl Chen
## Copyright (C) 2005 Hollis Blanchard

## This file is part of QuickyPix.

## QuickyPix is free software; you can redistribute it and/or modify it under
## the terms of the GNU General Public License as published by the Free
## Software Foundation; either version 2, or (at your option) any later
## version.

## QuickyPix is distributed in the hope that it will be useful, but WITHOUT
## ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
## FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
## more details.

## You should have received a copy of the GNU General Public License along
## with QuickyPix; see the file COPYING.  If not, write to the Free Software
## Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
## USA.

import sys
import os
import re
import time
import urllib
import mimetypes
import config

FULLSIZE = 'fullsize'

def frac_to_dec(fraction):
    numerator, denominator = str(fraction).split('/')
    return float(numerator) / float(denominator)

# returns tuple (x,y) for requested size.  (-1,-1) means fullsize.
def parse_size(arg):
    if arg == FULLSIZE:
        return -1, -1
    m = re.match('^([0-9]+)x([0-9]+)$', arg)
    if not m: not_found()
    return tuple(map(int, m.groups()))

def sized(size):
    if size == (-1,-1):
        return FULLSIZE
    else:
        return "%dx%d" %size

def get_mime_type(path):
    m = mimetypes.guess_type(path)[0]
    if not m:
        raise "Don't know mime type for %s!"%path
    return m

def rfc1123time(unixtime=None):
    """Returns time format preferred for Internet standards.

    Leeched from rfc822.formatdate().

    Sun, 06 Nov 1994 08:49:37 GMT  ; RFC 822, updated by RFC 1123

    According to RFC 1123, day and month names must always be in
    English.  If not for that, this code could use strftime().  It
    can't because strftime() honors the locale and could generate
    non-English names.

    """
    timeval = time.gmtime(unixtime or time.time())
    return "%s, %02d %s %04d %02d:%02d:%02d GMT" % (
            ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"][timeval[6]],
            timeval[2],
            ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
             "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"][timeval[1]-1],
                                timeval[0], timeval[3], timeval[4], timeval[5])

def cat(path, save_filename=None):
    '''Write a file to an HTTP client on stdout.

    save_filename, if specified, is the filename to suggest to the browser
    to save the file as.

    if save_filename == 1, then it is basename(path) iff config.download_flag.
    '''

    stat = os.stat(path)
    print "Content-Type: %s" %get_mime_type(path)
    print "Content-Length: %s" %stat.st_size
    if save_filename == 1:
        save_filename = config.download_flag and os.path.basename(path)
    if save_filename:
        print 'Content-Disposition: attachment; filename="%s"'%save_filename
    print "Last-Modified: %s" %rfc1123time(stat.st_mtime)
    print

    # sys.stdout.write(open(path,'rb').read())
    f = open(path,'rb')
    while True:
        d = f.read(4096)
        if not d: break
        sys.stdout.write(d)


def error_forbidden(data=''):
    print 'Status: 403 Forbidden'
    print 'Content-type: text/html\n'
    print '<h1>Forbidden :( %s</h1>'%data
    raise SystemExit

def not_found():
    print 'Status: 404 Not found'
    print 'Content-type: text/html\n'
    print '<h1>File not found :(</h1>'
    print 'I couldn\'t find <code>%s</code>.'%config.request_uri
    raise SystemExit

def redirect(url):
    print "Status: 302 Moved Temporarily"
    print 'Location: %s' %url
    print
    raise SystemExit

def relative_url(url):
    return "%s/%s"%(
        os.path.dirname(reconstruct_current_url()), url)

def redirect_rel(url):
    redirect(relative_url(url))

def reconstruct_root_url():
    if os.environ.get('HTTPS') == 'on':
        proto = 'https'
    else:
        proto = 'http'
    return '%s://%s%s' %(proto, os.environ.get('HTTP_HOST'), config.root)

def reconstruct_current_url():
    return reconstruct_root_url() + config.input_path

def escHtml(st):
    """Return an escaped version of an HTML string.
    """
    if not st: return st
    st = st.replace("&", "&amp;")
    st = st.replace("<", "&lt;")
    st = st.replace(">", "&gt;")
    st = st.replace('"', "&quot;")
    return st

escUrl = urllib.quote

def str2paragraphs(st):
    if not st: return st
    return '<p>%s</p>' % st.replace('\n\n','</p><p>')

def commify(number):
    return commify_('%d'%number)

def commify_(str):
    if len(str) <= 3:
        return str
    return commify_(str[:-3]) + ',' + str[-3:]


def xstat(filename):
    try:
        return os.stat(filename)
    except OSError:
        return None

def recent(stat):
    return stat and (time.time()-stat.st_mtime < config.RECENT_THRESHOLD)

def read_file(filename):
    try:
        return open(filename).read().strip()
    except:
        return ''

class AccessError(Exception):
    def __init__(self, filename):
        self.filename = filename

    def print_exit(self):
        print "Status: 403 forbidden"
        print "Content-Type: text/html"
        print
        print "<h1>Forbidden: %s</h1>" %self.filename
        raise SystemExit

def write_file(filename, data):
    try:
        open(filename,'w').write(data)
    except Exception, e:
        raise AccessError(filename)


def log(stuff):
    print >>open(config.LOG_PATH,'a'), time.strftime('[%Y-%m-%d %H:%M:%S]'), stuff

class MetadataValue(str):
    pass

class MetadataProperty:
    def __init__(self, filename, cname, default):
        self.filename = filename
        self.cname = cname
        self.cache = None
        self.default = default

    def get(self, obj):
        if self.cname in obj.__dict__:
            return obj.__dict__[self.cname]

        fn = config.ALBUMS_DIR + obj.fullpath + self.filename
        v = read_file(fn)

        if not v and self.default:
            # If there's a default (a function), evaluate it: caching it,
            # without writing to disk.
            v = self.default(obj)

        v = MetadataValue(v)
        v.stat = xstat(fn)
        v.recent = recent(v.stat)
        obj.__dict__[self.cname] = v
        return v

    def set(self, obj, value):
        value = value and value.strip() or ''
        old_value = self.get(obj)
        if old_value == value: return
        path = obj.fullpath
        fn = config.ALBUMS_DIR + path + self.filename
        if not path.startswith('/') or '/' in self.filename or '../' in fn:
            raise ValueError
        #del obj.__dict__[self.cname]
        value = MetadataValue(value)
        value.recent = True
        if not getattr(value,'nosave',False):
            obj.__dict__[self.cname] = value
            write_file(fn, value)
            log("Modifying %s: from %s to %s"%(fn, repr(old_value), repr(value)))

def metadata_property(filename, default=None):
    # Returns a property with getter and setter for reading and writing file.
    # Memoized.  Automatic strip()ing.  Value returned by getter also has a
    # 'recent' subproperty.
    cname = 'cache|%s'%filename
    mp = MetadataProperty(filename, cname, default)
    return property(mp.get, mp.set)

badchars = re.compile('[/\\*?\n\r|~;$^]')
# '&' no longer in badchars
def name_validate(rel_name):
    if (not rel_name or
        re.search(badchars,rel_name) or
        rel_name.startswith('.')):
        error_forbidden('16bbc225-e04e-47de-b074-99aad21831ac')

def skin_path_validate(name):
    if (not re.match('([a-zA-Z][a-zA-Z_.]*/)*[a-zA-Z][a-zA-Z_.]*[.](css|js)', name)):
        error_forbidden('a7149d71-cd46-49a2-a7c7-ed3ea6c7dd00')

def ensure_dir(fpath):
    dir = os.path.dirname(fpath)
    if not os.path.exists(dir):
        os.makedirs(dir)

def newer_than(f1, f2):
    # print >>sys.stderr, "## %s=%d, %s=%d" %(f1,os.stat(f1).st_mtime , f2,os.stat(f2).st_mtime)
    return os.stat(f1).st_mtime > os.stat(f2).st_mtime

def tdel(fpath):
    if os.spawnlp(os.P_WAIT, config.PATH_TDEL, 'tdel', '-d',
                  fpath):
        raise "Couldn't tdel %s"%fpath

def path_ext(path):
    if '.' in path:
        return path.split('.')[-1]
    else:
        return ''

def path_nonext(path):
    if '.' in path:
        return path[:path.rfind('.')]
    else:
        return path

def is_media(path):
    return (path_ext(path).lower() in config.MEDIA_TYPES)
